bwe/BandwidthEstimator.cpp

```diff
@@ -31,164 +43,243 @@ BandwidthEstimator::BandwidthEstimator(const Config& config)
         _previousReceiveTime = receiveTimeNs - 5 * utils::Time::ms;
+        _packetSize0 = packetSize;
     }
 
     const double tau = std::max(0.0, static_cast<double>(transmitTimeNs - _previousTransmitTime) / utils::Time::ms);
     const double observedDelay =
         static_cast<double>(static_cast<int64_t>(receiveTimeNs - transmitTimeNs - _baseClockOffset)) / utils::Time::ms;
 
-    const auto actualDelay = (observedDelay - _state(2));
-    const double congestionScale = analyseCongestion(actualDelay, packetSize, receiveTimeNs);
+    const auto expectedState = transitionState(packetSize, tau, _state);
+    const auto expectedDelay = predictAbsoluteDelay(expectedState);
+    _congestion.countDelays(observedDelay - expectedDelay);
 
-    _receiveBandwidth.update(packetSize * 8, receiveTimeNs);
-    // predict mean state
+    auto actualDelay = (observedDelay - _state(ClockOffset));
+    if (actualDelay < 0)
+    {
+        _state(QueuedBits) = 0; // queue must be empty before this packet
+        _state(ClockOffset) = observedDelay;
+        actualDelay = 0;
+        _packetSize0 = packetSize;
+    }
+    else if (actualDelay == 0 && packetSize > _packetSize0)
+    {
+        _packetSize0 = packetSize;
+    }
+    else if (observedDelay > expectedDelay && tau > 20.0 && expectedState(QueuedBits) < _config.mtu * 8 * 3)
+    {
+        // inter transmit interval larger than 20ms risk having radio shut down. adjust co up
+        _state(ClockOffset) += std::min(30.0, observedDelay - expectedDelay);
+        _packetSize0 = packetSize;
+    }
+
+    auto processNoise = _processNoise;
+    double measurementNoise = _config.measurementNoise;
+    calculateProcessNoise(transitionState(packetSize, tau, _state),
+        actualDelay,
+        observedDelay - expectedDelay,
+        packetSize,
+        receiveTimeNs,
+        processNoise,
+        measurementNoise);
+
+    measurementNoise *= analyseCongestion(actualDelay, packetSize, receiveTimeNs);
+
+    _receiveBitrate.update(packetSize * 8, receiveTimeNs);
+
+    // generate alternative current positions
     std::array<math::Matrix<double, 3>, SIGMA_POINTS> sigmaPoints;
-    generateSigmaPoints(_state, _covarianceP, _processNoise, sigmaPoints);
+    generateSigmaPoints(_state, _covarianceP, processNoise, sigmaPoints);
 
+    // calculate where we would have transitioned from alternative positions, and what the delay would be then
     std::array<double, SIGMA_POINTS> predictedDelays;
     for (size_t i = 0; i < sigmaPoints.size(); ++i)
     {
-        sigmaPoints[i] = transitionState(packetSize, tau, sigmaPoints[i], observedDelay);
-        predictedDelays[i] = predictDelay(sigmaPoints[i]);
+        sigmaPoints[i] = transitionState(packetSize, tau, sigmaPoints[i]);
+        predictedDelays[i] = predictAbsoluteDelay(sigmaPoints[i]);
     }
+    const double predictedMeanDelay = predictedDelays[0]; // delay of mean state, because 1/bw is non-linear
```

Comment:
I noticed that the mean delay from all sigma points is always offset upwards, because the transformation is non-linear. 
Using the delay from the current state is a better starting point for the variance calculation, though it bends ukf a bit.